KEP-26: Implementation

porridge opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kudobuilder/kudo/blob/master/CONTRIBUTING.md
2. Make sure you have added and ran the tests before submitting your PR
3. If the PR is unfinished, start it as a Draft PR: https://github.blog/2019-02-14-introducing-draft-pull-requests/
-->

**What this PR does / why we need it**:

Add a `--parameter-file` flag.
This is for [KEP-26](#1364)

TODO:
- [x] Documentation on kudo.dev/: https://github.com/kudobuilder/kudo.dev/pull/221

porridge commented on pkg/kudoctl/cmd/install.go:
```diff
@@ -53,6 +54,7 @@ func newInstallCmd(fs afero.Fs) *cobra.Command {
 
 	installCmd.Flags().StringVar(&options.InstanceName, "instance", "", "The Instance name. (defaults to Operator name appended with -instance)")
 	installCmd.Flags().StringArrayVarP(&parameters, "parameter", "p", nil, "The parameter name and value separated by '='")
+	installCmd.Flags().StringArrayVarP(&parameterFiles, "parameter-file", "P", nil, "YAML file with parameters")
```
The description explains the semantics of a single flag argument, and I made its wording consistent with the description for `--parameter`:
```
  -p, --parameter stringArray        The parameter name and value separated by '='
  -P, --parameter-file stringArray   YAML file with parameters
```

The `stringArray` is a hint that multiple such flags can be provided. It's not great UX, but native kubectl seems to be even worse in this area:
```
      --env=[]: Environment variables to set in the container
```

zen-dog replied:
So we allow multiple `--parameter-file`s to be passed? What are the rules then if both have the same key?

porridge commented on pkg/kudoctl/cmd/install/params.go:
```diff
@@ -4,13 +4,54 @@ import (
-func GetParameterMap(raw []string) (map[string]string, error) {
+// GetParameterMap takes a slice of parameter strings and a slice of parameter filenames as well as a filesystem,
+// parses parameters into a map of keys and values
+func GetParameterMap(raw []string, filePaths []string, fs afero.Fs) (map[string]string, error) {
```
- changed the signature to put `fs` first
- explained that values are marshalled in godoc
- perhaps `Marshal` would be better than `Wrap`, but I think changing the name is outside the scope of this PR, I'm merely using this pre-existing function
- I extracted the two key pieces from the function so that it's clear on first sight what it's doing
- I don't think it makes sense to pass the values to the original function, since the purpose of the original function is concerned with splitting key and value on the `=` sign, while the function which reads a parameter value file gets the key and value separately from the YAML parser.

porridge commented on pkg/kudoctl/cmd/install/params.go:
```diff
@@ -4,13 +4,54 @@ import (
+			case string:
+				valueType = v1beta1.StringValueType
+			default:
+				valueType = v1beta1.StringValueType
```
It *will* happen whenever a user provides a value of such type in the parameter value file. However since we do not have parameter type (not even parameter name!) information at this stage, we cannot do much about this other than marshal to string, which should DTRT in most cases anyway.

Note that this problem pre-dates this PR. The user can feed any kind of garbage parameter to the CLI and we will happily stuff it into the `Instance`.
I raised this [on slack](https://kubernetes.slack.com/archives/CG3HTFCMV/p1583845701167700) and the consensus so far seems to be to do validation in the API server layer. But I think you are well aware of this since you filed https://github.com/kudobuilder/kudo/issues/1411

porridge commented:
As for preventing combined `-p` and `-P` or preventing more than one `-P`, I strongly disagree.
- This is a very useful thing, if you can quickly iterate by overriding a single value without having to edit the whole file.
- This was explicitly [designed](https://github.com/kudobuilder/kudo/blob/master/keps/0026-reading-parameter-values-from-a-file.md#parameter-precedence-rules) and [discussed](https://github.com/kudobuilder/kudo/pull/1364#issuecomment-592456174) this way.
- This is a common pattern in other utilities, see for example `--env` and `--env-file` in `docker run`:
```bash
porridge@beczulka:~$ head ?.env
==> x.env <==
bla=filex

==> y.env <==
bla=filey
$ docker run -ti --env-file y.env --env-file x.env --env bla=cmd debian env|grep bla=
bla=cmd
$ docker run -ti --env-file x.env --env bla=cmd debian env|grep bla=
bla=cmd
$ docker run -ti --env-file x.env  debian env|grep bla=
bla=filex
$ docker run -ti --env-file x.env --env-file y.env debian env|grep bla=
bla=filey
$ docker run -ti --env-file y.env --env-file x.env debian env|grep bla=
bla=filex
$ 
```

PTAL @ken

kensipe commented:
@porridge you have really great logic and examples!   However I still disagree :) 

The value of multiple env and env-files may be useful in docker because it's UX is based on developers where a quick override could be useful.... the kudo experience is geared around operator experience.  The idea of a param file stems from 2 needs 1. the ability to express non-string types and 2. the ability to version a file of params that were used against a kudo operator.   I don't see the added complexity adding value.   I will review again setting that aside for now.

zen-dog commented on pkg/kudoctl/cmd/install/params.go:
```diff
@@ -4,13 +4,54 @@ import (
 	"errors"
 	"fmt"
 	"strings"
+
```
Now that param handling became more complex, maybe a new package is warranted? Wdyt about `cmd/params/parser.go` or smth. along those lines?

porridge commented:
@zen-dog 
> So we allow multiple `--parameter-files` to be passed?

Correct, see my [response to Ken's comment](https://github.com/kudobuilder/kudo/pull/1410#issuecomment-598051886)

> What are the rules then if both have the same key?

These are [documented on the site](https://github.com/kudobuilder/kudo.dev/pull/221) for users and [by the added unit test](https://github.com/kudobuilder/kudo/pull/1410/files#diff-3ad2d256de5bb9da73731663a0c76125R34) for developers :-)

porridge commented:
@kensipe the value from accepting both parameters and parameter values files is consistency with similar mechanism in other systems and programs, that sysadmins **do** expect. I know this because I happen to come from a sysadmin background. I think saying that UX is targeted at developers and not operators is a stretch, but that was just an example. Here is another [from systemd](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Environment).

I refactored the function to hopefully be less conflated :-) PTAL

@zen-dog I moved to another package, PTAL

zen-dog commented:
> @zen-dog I moved to another package, PTAL

👍 and I see how multiple `param-file`s also make sense.

zen-dog commented on pkg/kudoctl/cmd/params/parse.go:
```diff
@@ -0,0 +1,112 @@
+package params
```
I prefer file names that are nouns, and we already have multiple `parser.go` files in the project. Let's stay consistent.

porridge replied:
Done.
